### PR TITLE
Story/540 archiving restrictions model & config

### PR DESCRIPTION
**Pull request by PeterAlabaster**
To get an idea what the views look like:
image
image

Superuser/user who can access:
image
image

Restricted user:
image
image

Check is both on frontend and backend to block against malicious users de-activating warehouses via odoorpc

### CODE REVIEW

File: addons/udes_security/models/models.py [hunk 1]
```diff
@@ -7,9 +7,27 @@
 class BaseModel(models.AbstractModel):
     _inherit = "base"
 
+    @api.multi
+    def write(self, vals):
+        """Inherit write to add a check against res.archiving.restrictions
+        when the active field is written to.
+        """
+        ResArchivingRestrictions = self.env["res.archiving.restrictions"]
+        if "active" in vals:
```
Comment (PeterAlabaster) [inline addons/udes_security/models/models.py:16]: not sure if we also need to check that we are writing the opposite value to any of the records in `self` in case there are some wonky writes going on in certain models which update `"active" : True` when it is already set to True.
If there are and the user triggering those wonky writes doesn't have archiving permission, it will blow up.

File: addons/udes_security/models/res_archiving_restrictions.py [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+                "An Archiving Restriction already exists for this model. Please edit the existing one"
+            ),
+        )
+    ]
+
+    model_id = fields.Many2one(
+        "ir.model", string="Model to apply restriction to", required=True, index=True
+    )
+    name = fields.Char(string="Model Name", related="model_id.model", readonly=True)
+    group_ids = fields.Many2many(
```
Comment (PeterAlabaster) [inline addons/udes_security/models/res_archiving_restrictions.py:33]: self nit: maybe could call this `group_permitted_ids`

File: addons/udes_security/models/models.py (hunk 1 above)
Comment (PeterAlabaster) [reply addons/udes_security/models/models.py:16]: i was thinking something like this
```python
if "active" in vals:
    current_active = self.mapped("active")
    setting_active = vals.get("active")
    if len(current_active) > 1 or setting_active not in current_active:
        ResArchivingRestrictions._check_can_archive(self._name)
```

File: addons/udes_security/models/models.py [hunk 3]
```diff
@@ -7,9 +7,27 @@
...
+
     @api.model
     def fields_view_get(self, view_id=None, view_type="form", toolbar=False, submenu=False):
-        """Override to remove buttons/editable functionality for view only desktop users"""
+        """Override to remove buttons/editable functionality for view only desktop users
+        and take into account archiving restrictions for the given user"""
+
+        def _remove_active(doc):
+            """Remove any field or smart button related to the active field from the doc"""
+            for node in doc.xpath("//button[@icon='fa-archive']"):
```
Comment (PeterC10) [inline addons/udes_security/models/models.py:27]: We shouldn't be removing buttons based on what icon they have imo, unlikely be we could be unintentionally removing a button that is needed.

Could we change this to `@name='toggle_active'`? This is the core functionality for archiving in Odoo.

File: addons/udes_security/models/models.py [hunk 4]
```diff
@@ -74,6 +93,15 @@ def _set_fields_readonly(doc, res):
 
             res["arch"] = etree.tostring(doc)
             # Remove archive option in tree view and make calendar readonly
-            res["readonly_desktop_user"] = True
+            res["readonly_desktop_user"] = True  # Used in basic_view.js
+
+        can_archive_record = ResArchivingRestrictions._get_can_archive(model)
```
Comment (PeterC10) [inline addons/udes_security/models/models.py:98]: Can we `else` this with `if self.env.user.u_desktop_readonly and not ModelException.is_exempt(model):`?

If a user is readonly then they're not going to be able to archive anything anyway so the checks here aren't needed.

File: addons/udes_security/models/res_archiving_restrictions.py [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+    """
+
+    _name = "res.archiving.restrictions"
+    _description = "Record Archiving Restrictions"
+    _sql_constraints = [
+        (
+            "model_id_uniq",
+            "unique(model_id)",
+            _(
+                "An Archiving Restriction already exists for this model. Please edit the existing one"
```
Comment (PeterC10) [inline addons/udes_security/models/res_archiving_restrictions.py:24]: Super super super nit, feel free to resolve but I think the message should be set as one of the following:

`An Archiving Restriction already exists for this model, please edit the existing one`

or

`An Archiving Restriction already exists for this model. Please edit the existing one.`

or just:

`An Archiving Restriction already exists for this model`

File: addons/udes_security/models/res_archiving_restrictions.py (hunk 2 above)
Comment (PeterC10) [reply addons/udes_security/models/res_archiving_restrictions.py:33]: Yeah tricky one, as it's hard to work out straight away if `group_ids` means they are exempt or the only ones affected!

Personally I prefer `permitted_group_ids`. :)

File: addons/udes_security/models/res_archiving_restrictions.py [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+        )
+    ]
+
+    model_id = fields.Many2one(
+        "ir.model", string="Model to apply restriction to", required=True, index=True
+    )
+    name = fields.Char(string="Model Name", related="model_id.model", readonly=True)
+    group_ids = fields.Many2many(
+        "res.groups",
+        string="Groups permitted to archive & unarchive",
```
Comment (PeterC10) [inline addons/udes_security/models/res_archiving_restrictions.py:35]: Nit: `Groups Permitted to Archive & Unarchive`

File: addons/udes_security/models/res_archiving_restrictions.py [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+        help="""
+        If no groups are configured, ALL users will be prevented from archiving (except superuser)
+        If the user is in any of the configured groups, they will be permitted access
+        If no restriction on a model is intended, then do not create a restriction record for that model
+        """,
+    )
+
+    def _get_can_archive(self, model_name):
+        """
+        Test against the model to see if our user is
```
Comment (PeterC10) [inline addons/udes_security/models/res_archiving_restrictions.py:46]: Suggestion: `Check against the model to see if active user is permitted to archive/unarchive the record.`

"Test" and "our user" read a bit odd to me.

File: addons/udes_security/models/res_archiving_restrictions.py [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+    def _check_can_archive(self, model_name):
+        """Raises a warning if the logged in user does not meet
+        the group requirements to archive/unarchive a record
+
+        :param: model_name: str() model _name i.e "stock.picking.type"
+        """
+        if not self._get_can_archive(model_name):
+            raise ValidationError(
+                _(
+                    f"Cannot archive/unarchive record due user not meeting restrictions in place on {model_name}"
```
Comment (PeterC10) [inline addons/udes_security/models/res_archiving_restrictions.py:79]: Suggestion: As this will be a user facing error if it ever comes up (which it shouldn't) I'd suggest rewording it slightly, e.g.:

`You do not have permission to archive/unarchive this record.`

File: addons/udes_security/models/res_archiving_restrictions.py [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+        return bool(groups_on_user_and_restriction)
+
+    def _check_can_archive(self, model_name):
+        """Raises a warning if the logged in user does not meet
+        the group requirements to archive/unarchive a record
+
+        :param: model_name: str() model _name i.e "stock.picking.type"
+        """
+        if not self._get_can_archive(model_name):
+            raise ValidationError(
```
Comment (PeterC10) [inline addons/udes_security/models/res_archiving_restrictions.py:77]: Suggestion: Log this, recording which user tried to archive the record.

File: addons/udes_security/security/ir.model.access.csv [hunk 10]
```diff
@@ -1,3 +1,4 @@
 id,name,model_id:id,group_id:id,perm_read,perm_write,perm_create,perm_unlink
 udes_allowed_file_type,udes_allowed_file_type,model_udes_allowed_file_type,base.group_user,1,0,0,0
 access_udes_readonly_desktop_user_model_exception,access_udes_readonly_desktop_user_model_exception,model_udes_readonly_desktop_user_model_exception,base.group_user,1,0,0,0
+access_res_archiving_restrictions,access_res_archiving_restrictions,model_res_archiving_restrictions,base.group_no_one,1,1,1,1
```
Comment (PeterC10) [inline addons/udes_security/security/ir.model.access.csv:4]: Suggestion: Add a readonly permission for all users for safety, like the readonly model exception.

File: addons/udes_security/models/res_archiving_restrictions.py [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+
+    If an Archiving Restriction is created for a model, the group_ids set will determine
+    which groups have the facility to archive and unarchive records on the model.
+
+    This includes backend writes to the `active` field, as well as hiding
+    the Archive/Unarchive action from the model, and attempting to remove active
+    fields & smart buttons from form views.
+    """
+
+    _name = "res.archiving.restrictions"
```
Comment (PeterC10) [inline addons/udes_security/models/res_archiving_restrictions.py:17]: Usually the names of records are singular, e.g. `res.archiving.restriction`.

I know Odoo is inconsistent with this itself, but I think it should still be singular here, as we have one record per model that is restricted.

File: addons/udes_security/models/res_archiving_restrictions.py [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+            "model_id_uniq",
+            "unique(model_id)",
+            _(
+                "An Archiving Restriction already exists for this model. Please edit the existing one"
+            ),
+        )
+    ]
+
+    model_id = fields.Many2one(
+        "ir.model", string="Model to apply restriction to", required=True, index=True
```
Comment (PeterC10) [inline addons/udes_security/models/res_archiving_restrictions.py:30]: We don't need to the string so long here, just `Model` will do. We can use this as the help text though. :)

File: addons/udes_security/models/models.py [hunk 13]
```diff
@@ -74,6 +93,15 @@ def _set_fields_readonly(doc, res):
...
+        if not can_archive_record:
+            doc = etree.XML(res["arch"])
+            if doc.tag == "form":
```
Comment (PeterC10) [inline addons/udes_security/models/models.py:101]: Q: What about the archive option from the tree view when you select some records?

File: addons/udes_security/models/models.py (hunk 1 above)
Comment (PeterC10) [reply addons/udes_security/models/models.py:16]: Will need to try out this scenario, I'm sure in some cases I've seen Odoo filtering out writes that are writing the existing value, but might be bad memory or wishful thinking on my part...

File: addons/udes_security/models/res_archiving_restrictions.py (hunk 6 above)
Comment (PeterC10) [reply addons/udes_security/models/res_archiving_restrictions.py:35]: Thinking about this more... maybe just `Permitted Groups` as the string, and then `Groups permitted to archive & unarchive` as the help text?

File: addons/udes_security/models/models.py (hunk 1 above)
Comment (ws-mps) [reply addons/udes_security/models/models.py:16]: I don't think it does unfortunately. But it doesn't matter, let's keep it simple, if the user cannot change the value shouldn't be in vals, no need to only check restriction if the value is different.

File: addons/udes_security/models/res_archiving_restrictions.py [hunk 14]
```diff
@@ -0,0 +1,82 @@
...
+
+        Can archive if user has _any_ groups defined on the restriction,
+        or no restriction is defined
+
+        :param: model_name: str() model _name i.e "stock.picking.type"
+        :returns: bool()
+        """
+        ResArchivingRestrictions = self.env["res.archiving.restrictions"]
+
+        if self.env.user.id == SUPERUSER_ID:
```
Comment (adamrpatrick) [inline addons/udes_security/models/res_archiving_restrictions.py:57]: What about debug/ trusted users?

Comment (PeterC10) [reply addons/udes_security/models/res_archiving_restrictions.py:57]: Trusted user is added here: https://github.com/unipartdigital/udes-open/blob/story/540/addons/udes_stock/data/res_archiving_restrictions.xml

But I think this is the wrong way around, should be debug user and not trusted user.

File: addons/udes_security/models/res_archiving_restrictions.py (hunk 9 above)
Comment (PeterAlabaster) [reply addons/udes_security/models/res_archiving_restrictions.py:77]: Just to confirm messaging:
Frontend (user message):
```
You do not have permission to archive/unarchive this record.
```
Backend  log message:
```
2021-12-21 09:40:17,449 412241 WARNING mvp odoo.addons.udes_security.models.res_archiving_restrictions: User 'bob', ID 6 blocked attempt at archiving on 'stock.picking.type'.
```

File: addons/udes_security/models/res_archiving_restrictions.py (hunk 11 above)
Comment (PeterAlabaster) [reply addons/udes_security/models/res_archiving_restrictions.py:17]: Yeah, i wondered about this because even I was being inconsistent with naming and `res.archiving.restriction` seemed slightly better, but didn't want to waste time doing it if it was going to be scrapped anyway. Will get it changed (Y)

File: addons/udes_security/models/res_archiving_restrictions.py [hunk 15]
```diff
@@ -0,0 +1,82 @@
...
+            _(
+                "An Archiving Restriction already exists for this model. Please edit the existing one"
+            ),
+        )
+    ]
+
+    model_id = fields.Many2one(
+        "ir.model", string="Model to apply restriction to", required=True, index=True
+    )
+    name = fields.Char(string="Model Name", related="model_id.model", readonly=True)
```
[the remark this answers is not in the document]
Comment (PeterAlabaster) [reply addons/udes_security/models/res_archiving_restrictions.py:32]: Solely used for nice displayname in the breadcrumb

File: addons/udes_security/models/models.py (hunk 13 above)
Comment (PeterAlabaster) [reply addons/udes_security/models/models.py:101]: That is handled by `cannot_archive_record` that's shoved into res, and picked up in BasicView.js

File: addons/udes_security/models/res_archiving_restrictions.py (hunk 14 above)
Comment (PeterAlabaster) [reply addons/udes_security/models/res_archiving_restrictions.py:57]: Following principal with rest of Odoo, superuser can do anything, regardless of individual permissions set up.

@PeterC10 is right and trusted user is configured as the only group permitted access to picking type and warehouse, but re-reading the story, it says only debug user, not trusted user.

I'll clarify w/ product whether they want just debug user, or trusted user, or both

File: addons/udes_security/models/res_archiving_restrictions.py (hunk 9 above)
Comment (PeterC10) [reply addons/udes_security/models/res_archiving_restrictions.py:77]: Yep that looks good. :)

File: addons/udes_security/models/res_archiving_restrictions.py (hunk 14 above)
Comment (PeterAlabaster) [reply addons/udes_security/models/res_archiving_restrictions.py:57]: Confirmed it is only debug users, so will get it changed in the data file in udes_stock

**Review by adamrpatrick — APPROVED** [review]

**Review by ws-mps — APPROVED** [review]

**Review by adamrpatrick — APPROVED** [review]
